### DISCUSSION TITLE
Lucidjs: DatabaseQueryBuilder macroable

### DISCUSSION THREAD

**Discussion by ad-momo** [adonisjs/core#4791] [Help]
Hi 👋,

**Description**

In Adonis v5, the DatabaseQueryBuilder can be macroable [by following the instructions](https://v5-docs.adonisjs.com/guides/database/query-builder#extending-query-builders)

I want to do this in Adonis v6 but there are no longer any instructions in the Adonis/Lucid documentation. So I reproduce the same logic as Vine customization but I have a typescript error :
> Argument of type '"whereComplexClause"' is not assignable to parameter of type 'keyof DatabaseQueryBuilder'.ts(2345)

DatabaseQueryBuilder is still macroable?

**pnpm list "@adonisjs/*"**

dependencies:
@adonisjs/auth 9.2.3
@adonisjs/bouncer 3.1.3
@adonisjs/core 6.12.1
@adonisjs/cors 2.2.1
@adonisjs/drive 3.2.0
@adonisjs/lucid 21.1.0
@adonisjs/session 7.5.0

devDependencies:
@adonisjs/assembler 7.7.0
@adonisjs/eslint-config 1.3.0
@adonisjs/prettier-config 1.3.0
@adonisjs/tsconfig 1.3.0

**Reproduce**

```ts
// start/database/query_builder.ts

import { DatabaseQueryBuilder } from '@adonisjs/lucid/database'
import {
  ChainableContract,
  Dictionary,
  ExcutableQueryBuilderContract,
} from '@adonisjs/lucid/types/querybuilder'

declare module '@adonisjs/lucid/database' {
  interface DatabaseQueryBuilderContract<Result = Dictionary<any, string>>
    extends ChainableContract,
      ExcutableQueryBuilderContract<Result[]> {
    whereComplexClause(data: Data): this
  }
}

DatabaseQueryBuilder.macro(
  'whereComplexClause', // 👈 Here I have TS error : "Argument of type '"whereComplexClause"' is not assignable to parameter of type 'keyof DatabaseQueryBuilder'.ts(2345)"
  function (this: DatabaseQueryBuilder, data: Data) {
    return this.whereILike(...)
  }
)
```

```ts
// adonisrc.ts
...
  preloads: [
    () => import('#start/routes'),
    () => import('#start/kernel'),
    () => import('#start/database/query_builder'),
  ],
...
```

Thanks 😊

**Comment by douglasdnzcarvalho** [comment]
I'm migrating a considerably large application (+- 300 files) to v6 and I'm stuck with this problem (among many others)... have you found any solution other than refactoring all the places where the macro is used?

**Reply by ad-momo** [reply]
Hello :wave:,

Yes, in my case, with query callback :

```ts
// query_utils.ts

export function complexClauseCallback(data: Data): QueryCallback<ChainableContract> {
  return (query: ChainableContract) => {
    query.whereILike(...)
  }
}

```

```ts
// service.ts
...

db.query()
// select..
.where(complexClauseCallback(data))

...
```
